File: thinking-categorically/domain-and-codomain/04-higher-order functions/higher-order-functions.cs
List<int> numbers = new List<int> { 1, 2, 3, 4, 5 };

// Select is map: Func<a, b> -> IEnumerable<a> -> IEnumerable<b>
var doubled = numbers.Select(x => x * 2);

// Where is filter: Func<a, bool> -> IEnumerable<a> -> IEnumerable<a>
var evens = numbers.Where(x => x % 2 == 0);

// Aggregate is reduce/fold
var sum = numbers.Aggregate(0, (acc, x) => acc + x);

// Higher-order function that returns a function
Func<int, Func<int, int>> CreateAdder = x => y => x + y;

var addTen = CreateAdder(10);
var result = addTen(5); // 15

Console.WriteLine(result);

File: thinking-categorically/domain-and-codomain/04-higher-order functions/higher-order-functions.ts
// Higher-order functions are everywhere in JavaScript
const numbers: number[] = [1, 2, 3, 4, 5];

// map: (a -> b) -> Array<a> -> Array<b>
const doubled: number[] = numbers.map(x => x * 2);

// filter: (a -> boolean) -> Array<a> -> Array<a>
const evens: number[] = numbers.filter(x => x % 2 === 0);

// reduce: (accumulator -> a -> accumulator) -> accumulator -> Array<a> -> accumulator
const sum: number = numbers.reduce((acc, x) => acc + x, 0);

// Function factory - returns customized functions
function createMultiplier(factor: number): (x: number) => number {
  return (x: number) => x * factor;
}

const triple = createMultiplier(3);
const result = triple(4); // 12

console.log(result);
